pkg/gardenlet/controller/federatedseed/extensions/shootstate_control.go

```diff
@@ -61,75 +73,87 @@ func (s *shootStateControl) createShootStateSyncReconcileFunc(kind string, objec
-				return reconcile.Result{}, nil
+		// Delete resources which are no longer present in the extension state from the ShootState.Spec.Resources list
+		for _, resource := range currentResources {
+			if gardencorev1beta1helper.GetResourceByName(newResources, resource.Name) == nil {
+				updateShootStateResourceData(shootState, &resource.ResourceRef, nil)
 			}
-			return reconcile.Result{}, fmt.Errorf("could not get cluster with name %s : %v", clusterName, err)
 		}
 
-		shoot, err := s.shootRetriever.FromCluster(cluster)
+		resourcesToUpdate, err := s.getResourcesToUpdate(ctx, shootState, extensionObj.GetNamespace(), newResources)
 		if err != nil {
 			return reconcile.Result{}, err
 		}
+		for _, resourceToUpdate := range resourcesToUpdate {
+			updateShootStateResourceData(shootState, &resourceToUpdate.CrossVersionObjectReference, &resourceToUpdate.Data)
+		}
 
-		name := extensionObject.GetName()
-		purpose := extensionObject.GetExtensionSpec().GetExtensionPurpose()
-		state := extensionObject.GetExtensionStatus().GetState()
-		resources := extensionObject.GetExtensionStatus().GetResources()
-
-		shootState := &gardencorev1alpha1.ShootState{ObjectMeta: metav1.ObjectMeta{Name: shoot.Name, Namespace: shoot.Namespace}}
-		if _, err := controllerutil.CreateOrUpdate(ctx, s.k8sGardenClient.Client(), shootState, func() error {
-			// Delete resource data for resources that are no longer present in the extension state
-			_, currentResources := getShootStateExtensionState(shootState, kind, &name, purpose)
-			for _, currentResource := range currentResources {
-				if gardencorev1beta1helper.GetResourceByName(resources, currentResource.Name) == nil {
-					updateShootStateResourceData(shootState, &currentResource.ResourceRef, nil)
-				}
-			}
+		shouldUpdateExtensionData := !apiequality.Semantic.DeepEqual(newState, currentState) || !apiequality.Semantic.DeepEqual(newResources, currentResources)
+		if shouldUpdateExtensionData {
+			updateShootStateExtensionStateAndResources(shootState, kind, &name, purpose, newState, newResources)
+		}
 
-			// Update the extension state
-			updateShootStateExtensionState(shootState, kind, &name, purpose, state, resources)
-
-			// Update resource data for resources that are still present in the extension state
-			for _, resource := range resources {
-				obj, err := utils.GetObjectByRef(ctx, s.seedClient.Client(), &resource.ResourceRef, extensionObject.GetNamespace())
-				if err != nil {
-					return err
-				}
-				if obj == nil {
-					return fmt.Errorf("object not found %v", resource.ResourceRef)
-				}
-				raw := &runtime.RawExtension{}
-				if err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj, raw); err != nil {
-					return err
-				}
-				updateShootStateResourceData(shootState, &resource.ResourceRef, raw)
+		if shouldUpdateExtensionData || len(resourcesToUpdate) > 0 {
+			if err := s.k8sGardenClient.Client().Patch(ctx, shootState, client.MergeFrom(shootStateCopy)); err != nil {
```

Comment:
This patch call uses a `json-merge` patch. So the `mergeKeys` you added to the API, won't be leveraged.
Instead, the entire list of resources will be replaced.
We either need a `strategic-merge` patch or a `json-merge` patch with optimistic locking here.